### DISCUSSION TITLE
[UEFI Ipxe] After a boot failure, the machine will automatically restart after 5 minutes in the setup interface

### DISCUSSION THREAD

**Discussion by Feng-zero** [ipxe/ipxe#1133] [Q&A] [answered]
BIOS booted up and loaded the optionROM from NIC. In the Save&exit interface, I entered the IPxe boot, but failed to boot(Caused by server configuration issues) and returned to the BIOS setup interface. I operated on the interface for 5 minutes, but the watchdog triggered the action of restarting the machine. Is this a normal mechanism?
Is it set by: https://github.com/ipxe/ipxe/blob/master/src/interface/efi/efi_watchdog.c ?

**Comment by NiKiZe** [comment]
iPXE should initialize the watchdog timer yes.
But it should also shutdown on exit, see comment in https://github.com/ipxe/ipxe/blob/master/src/interface/efi/efi_watchdog.c#L95

**Reply by Feng-zero** [reply]
In my testing scenario, the BIOS did not exit, but instead returned to the setup interface due to an IPxe boot failure. So the ExitBootServices() interface of optionrom was not called, but the 5-minute watchdog is still running.

**Reply by Feng-zero** [reply]
![image](https://github.com/ipxe/ipxe/assets/57554575/6aae3757-4b51-4bfe-a3c5-ced0bc01f5f1)
I enter ipxe boot in Boot Override.

**Comment by mcb30** [comment, marked as answer]
> BIOS booted up and loaded the optionROM from NIC. In the Save&exit interface, I entered the IPxe boot, but failed to boot(Caused by server configuration issues) and returned to the BIOS setup interface. I operated on the interface for 5 minutes, but the watchdog triggered the action of restarting the machine. Is this a normal mechanism?

After exiting from iPXE, it is the responsibility of the platform firmware to either cancel the watchdog timer, or to periodically reset the timer.  You can see the relevant code in EDK2 at https://github.com/tianocore/edk2/blob/master/MdeModulePkg/Library/UefiBootManagerLib/BmLoadOption.c#L1485-L1502

> Is it set by: https://github.com/ipxe/ipxe/blob/master/src/interface/efi/efi_watchdog.c ?

iPXE will periodically reset the watchdog timer while iPXE itself is running, in order to prevent the system from rebooting after five minutes spent inside iPXE (e.g. waiting at the iPXE shell prompt for user interaction).

It sounds as though your platform firmware may have a bug in that it fails to either cancel the watchdog timer or to periodically reset the watchdog timer.  (Alternatively, it's possible that your firmware vendor made a deliberate choice to treat five minutes of inactivity as indicating that the system ought to be rebooted.)

**Reply by Feng-zero** [reply]
Thanks for your answers. My firmware should be problematic, thank you very much!
